framework/sandstone.cpp

```diff
@@ -870,6 +895,26 @@ static void *thread_runner(void *arg)
                 // no rethrow
             }
         }
+
+        long double get_effective_freq_mhz(void) const
+        {
+            // For when we get bogus values (e.g. running unprivileged)
+            if (mperf_bef >= mperf_aft || aperf_bef >= aperf_aft || tsc_bef >= tsc_aft) {
```

Comment:
As opposed to doing this, please implement error handling. `(read|write)_msr()` returns `false` if the value cannot be read or written.

Reply:
But the values could unexpectedly roll-back, especially if we're virtualised, so range checking is still necessary.

Reply:
I'll leave it as-is then, unless folks have/want a better way of doing this.